### PR TITLE
Add type coercions in join criteria

### LINKED ISSUE #2419: Type coercions in the JOIN ON clause broken
The following query fails: 
```
SELECT *
FROM
 (VALUES 1.0) a(x)
 JOIN 
 (VALUES 2) b
 ON a.x = 1
```

### PR DESCRIPTION
**Pull request by yuananf**
Fix #2419 
@cberner Can you take a look at this?

### CODE REVIEW

**Comment by cberner** [comment]
I think the tests are broken. Can you look at those?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 1]
```diff
@@ -290,6 +290,11 @@ public void addTypes(IdentityHashMap<Expression, Type> types)
         this.types.putAll(types);
     }
 
+    public void replaceType(Expression expression, Type type)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I don't think you need to replace the type. Just adding the coercion should be enough

**Comment by cberner** [comment]
Assign back to me when you've address the comments

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 1 above)
Comment (yuananf) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I thought so too.
But if I remove this line, the non-equi join criteria coercions will not work, and throw exception.

In function appendProjections in RelantionPlanner, 
Symbol symbol = symbolAllocator.newSymbol(expression, analysis.getType(expression));
If I dont's explicity replace the types map, analysis.getType(expression) will still get the origin type of expression but not the coercion type.

Any idea about this?

**Comment by yuananf** [comment]
I will fix the tests and update it.

**Comment by yuananf** [comment]
Tests broken are fixed now.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 1 above)
Comment (yuananf) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I can remove this, but the non-equi join criteria coercions should be fixd in another way.

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: Could you add a test case that reproduces that? I tried removing `replaceType()` and all the tests passed

Comment (yuananf) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I got a problem here, I added a test in my local code,
assertQuery(
                "SELECT * FROM (VALUES (1.0)) x (a) JOIN (VALUES (3)) y (a) ON y.a > x.a",
                "VALUES (1.0, 3)"
        );
I remove replaceType() in my code and it passed the test, but when I run it in my cluster, this SQL throws exception.

**Comment by yuananf** [comment]
Rebased and change to correct test case.
When remove replaceType() all the tests passed in mvn test, but non-equi join test sql fails in real presto cluster. Wondering why this happens @cberner

**Comment by cberner** [comment]
That's quite strange. What's the stack trace?

**Comment by yuananf** [comment]
Query 20150326_024028_00001_u4rur failed: java.lang.VerifyError: Bad type on operand stack
 Exception Details:
 Location:
 com/facebook/presto/$gen/PageProcessor_130.filter(Lcom/facebook/presto/spi/ConnectorSession;Lcom/facebook/presto/spi/block/Block;Lcom/facebook/presto/spi/block/Block;I)Z @92: invokedynamic
 Reason:
 Type long_2nd (current frame, stack[3]) is not assignable to double_2nd

**Comment by cberner** [comment]
Hmm that looks like a bug in the expression compiler. I've seen other reports of that too, will try to take a look when I have time. Or if you track it down, please submit a PR!

**Comment by cberner** [comment]
I think this will fix the bytecode error. Can you try that out, and try removing the replaceType again?

https://github.com/facebook/presto/pull/2614

**Comment by yuananf** [comment]
I tried #2614 , the equi-join coercion works well now without my fix.

But unfortunately, the non-equi join coercion still breaks. The same error just like before.

**Comment by cberner** [comment]
I see. The problem is likely that the coercions aren't being set correctly, possibly because a new expression was created somewhere. It's not safe to create new Expressions during analysis because Presto relies on a number of `IdentityHashMap`s to tag the expressions with information. If you have time to debug it, that'd be great! Otherwise, I'll try to look into it later in the week

**Comment by yuananf** [comment]
Ok, let me try.

**Comment by yuananf** [comment]
opened a new pull request #2631 
will dig into the non-equi join problem.

**Pull request closed without merging**